### PR TITLE
Add archive command to zip a sketch and its files

### PR DESCRIPTION
**Pull request by silvanocerza**
**Please check if the PR fulfills these requirements**

- [x] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [x] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?**
New CLI command and new gRPC interface.

- **What is the current behavior?**
User can't create an archive of a Sketch from the CLI.
gRPC consumer doesn't have an interface to archive a Sketch.

* **What is the new behavior?**
User can create a `zip` file containing all files from a Sketch, `build` directory is only included if explicitly specified with the `--include-build-dir` flag.

Some possible examples on how to call the command from the CLI:

```
arduino-cli archive
arduino-cli archive .
arduino-cli archive . <some_folder>
arduino-cli archive . <some_folder>/MySketchArchive.zip
arduino-cli archive <path_to_sketch>
arduino-cli archive <path_to_sketch> <some_folder>
arduino-cli archive <path_to_sketch> <some_folder>/MySketchArchive.zip
```

- **Does this PR introduce a breaking change?**
No.

* **Other information**:
None.

---

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

### CODE REVIEW

**Comment by silvanocerza** [comment]
Links check is failing again because of `https://codecov.io/gh/arduino/arduino-cli`, it seems to work finely on the browser for me though. 
The rest of the tests are finally working too. :tada:

File: commands/instances.go [hunk 1]
```diff
@@ -702,3 +707,151 @@ func LoadSketch(ctx context.Context, req *rpc.LoadSketchReq) (*rpc.LoadSketchRes
 		AdditionalFiles:  additionalFiles,
 	}, nil
 }
+
+// ArchiveSketch FIXMEDOC
+func ArchiveSketch(ctx context.Context, req *rpc.ArchiveSketchReq) (*rpc.ArchiveSketchResp, error) {
```
Comment (cmaglie) [inline commands/instances.go:712]: Other commands are in their corresponding `commands/xxxx` subfolder that mimics the `cli/xxxx`, In this case this command should be inside `commands/archive`.

File: commands/instances.go [hunk 2]
```diff
@@ -702,3 +707,151 @@ func LoadSketch(ctx context.Context, req *rpc.LoadSketchReq) (*rpc.LoadSketchRes
...
+		archivePath = archivePath.Join(sketchName + ".zip")
+	} else if archivePath.Ext() == "" {
+		archivePath = paths.New(archivePath.String() + ".zip")
+	}
+
+	if archivePath.Exist() {
+		return nil, fmt.Errorf("archive already exists")
+	}
+
+	archive, err := os.Create(archivePath.Clean().String())
```
Comment (cmaglie) [inline commands/instances.go:760]:
```suggestion
	archive, err := archivePath.Create()
```

File: commands/instances.go [hunk 3]
```diff
@@ -702,3 +707,151 @@ func LoadSketch(ctx context.Context, req *rpc.LoadSketchReq) (*rpc.LoadSketchRes
...
+			sketchFiles = append(sketchFiles, files...)
+		}
+	}
+	finalFiles := paths.PathList{}
+	for _, f := range sketchFiles {
+		if f.IsNotDir() {
+			finalFiles = append(finalFiles, f)
+		}
+	}
+	return finalFiles, nil
```
Comment (cmaglie) [inline commands/instances.go:821]:
```suggestion
	sketchFiles.FilterOutDirs()
	return sketchFiles, nil
```

File: rpc/commands/commands.proto [hunk 4]
```diff
@@ -221,10 +224,10 @@ message UpgradeReq {
...
 
 message UpgradeResp {
-  // Progress of the downloads of the platforms and libraries files.
-  DownloadProgress progress = 1;
-  // Description of the current stage of the upgrade.
-  TaskProgress task_progress = 2;
+	// Progress of the downloads of the platforms and libraries files.
+	DownloadProgress progress = 1;
+	// Description of the current stage of the upgrade.
+	TaskProgress task_progress = 2;
```
Comment (cmaglie) [inline rpc/commands/commands.proto:230]: unrelated change

File: test/conftest.py [hunk 5]
```diff
@@ -84,18 +85,29 @@ def run_command(pytestconfig, data_dir, downloads_dir, working_dir):
...
-    cli_path = os.path.join(str(pytestconfig.rootdir), "..", "arduino-cli")
+
+    cli_path = Path(pytestconfig.rootdir).parent / "arduino-cli"
     env = {
         "ARDUINO_DATA_DIR": data_dir,
         "ARDUINO_DOWNLOADS_DIR": downloads_dir,
         "ARDUINO_SKETCHBOOK_DIR": data_dir,
     }
-    os.makedirs(os.path.join(data_dir, "packages"))
+    (Path(data_dir) / "packages").mkdir()
```
Comment (cmaglie) [inline test/conftest.py:95]: wow ultra-fancy use of `/` operator :-D

File: test/test_archive.py [hunk 6]
```diff
@@ -0,0 +1,1226 @@
...
+    assert "sketch_simple/old.pde" in archive_files
+    assert "sketch_simple/other.ino" in archive_files
+    assert "sketch_simple/s_file.S" in archive_files
+    assert "sketch_simple/sketch_simple.ino" in archive_files
+    assert "sketch_simple/src/helper.h" in archive_files
+    assert "sketch_simple/build/adafruit.samd.adafruit_feather_m0/sketch_simple.ino.hex" in archive_files
+    assert "sketch_simple/build/adafruit.samd.adafruit_feather_m0/sketch_simple.ino.map" in archive_files
+    assert "sketch_simple/build/arduino.avr.uno/sketch_simple.ino.eep" in archive_files
+    assert "sketch_simple/build/arduino.avr.uno/sketch_simple.ino.hex" in archive_files
+    assert "sketch_simple/build/arduino.avr.uno/sketch_simple.ino.with_bootloader.hex" in archive_files
```
Comment (cmaglie) [inline test/test_archive.py:1224]: This is always the same, maybe it's more convenient to put these checks in a function?

File: commands/instances.go [hunk 7]
```diff
@@ -702,3 +707,151 @@ func LoadSketch(ctx context.Context, req *rpc.LoadSketchReq) (*rpc.LoadSketchRes
...
+			return nil, fmt.Errorf("Error adding file to archive: %v", err)
+		}
+	}
+
+	return &rpc.ArchiveSketchResp{}, nil
+}
+
+// Recursively retrieves all files in the sketch folder
+func getSketchContent(sketchFolder *paths.Path) (paths.PathList, error) {
+	sketchFiles, err := sketchFolder.ReadDir()
```
Comment (cmaglie) [inline commands/instances.go:801]: Maybe we could simplify the whole function by implementing a `ReadDirRecursive()` in go-paths-helper?

**Review by cmaglie — CHANGES_REQUESTED** [review]
I've made some suggestions to slightly simplify here and there.

I'm wondering if the case:

`arduino-cli archive . /path/to/some/dir/name`

should automatically infer that we mean `/path/to/some/dir/name.zip` when the `name` folder doesn't exists. Personally, if I don't specify explicitly `.zip` I would expect a `/path/to/some/dir/name: directory does not exists`, but I'll leave to @ubidefeo the last word on this.

**Comment by cmaglie** [comment]
Nice work BTW, especially on the tests!

File: commands/instances.go (hunk 7 above)
Comment (silvanocerza) [reply commands/instances.go:801]: Oh yeah! Didn't think about it.

File: test/test_archive.py (hunk 6 above)
Comment (silvanocerza) [reply test/test_archive.py:1224]: Yep, will do.

File: commands/instances.go (hunk 2 above)
Comment (silvanocerza) [reply commands/instances.go:760]: That's how it works right now, if it's a directory it saves inside that dir a `<sketch_name>.zip`, if it doesn't exists `.zip` is appended to the path, after all this it checks if there's a already an existing archive with that name.

**Review by cmaglie — APPROVED** [review]
:+1:
